model_analyzer/analyzer.py

```diff
@@ -324,16 +340,18 @@ def _create_inference_output_table(self, title, aggregation_tag='Max'):
         # Create headers
         table_headers = self._param_inference_headers[:]
         for metric in self._monitoring_metrics:
-            if metric not in self.dcgm_tags:
+            if metric in self.perf_metrics:
```

Comment:
Let's not do this. This is a double search here. You are looping through the list of metrics and then looping through a second list for a potential O(n^2) implementation. This is bad algorithmically. 

Ideally, every metric should know how to print out the headers, you do not need to distinguish anything. For example, if``aggregation_tag is needed by cpu_metrics, then it should added in the `header()` method for cpu_metrics derivation. 

Try to re-write to avoid the if-else all together.

Reply:
This was done to preserve the order of the metrics in the original monitoring_metrics list, as well as to leave out the dcgm metrics from this table. What I can do is change it into 1 if statement just checking that its not a DCGM metric. Then I can convert the metric lists into sets, giving us O(1) lookup.

Reply:
O(1) is better, but the point is to avoid RTTI (runtime type identification) all together. Anytime you see code that adds special behavior based on a "type" you are committing a code smell. Instead, you should derive child classes for each type that have the same method, so the caller doesn't need to know. There are several Design Patterns that can better address the situation above that are cleaner.

https://refactoring.guru/replace-type-code-with-subclasses

Reply:
I think I can move the creation of the metrics tables into the MetricsManager. It will look more like a visitor pattern then.